Sources/Build/llbuild.swift

```diff
@@ -117,15 +117,72 @@ public struct LLBuildManifestGenerator {
             stream <<< "  " <<< Format.asJSON(target.name)
             stream <<< ": " <<< Format.asJSON(target.outputs.values) <<< "\n"
         }
+        
+        stream <<< "  " <<< Format.asJSON("regenerate")
+        stream <<< ": " <<< Format.asJSON([plan.buildParameters.regenerateManifestToken.asString])
+        stream <<< "\n"
+
         stream <<< "default: " <<< Format.asJSON(targets.main.name) <<< "\n"
+
+        // Add manifest regeneration directory nodes as directory structure.
+        let manifestRegenerationInputs = self.manifestRegenerationInputs()
+        if let manifestRegenerationInputs = manifestRegenerationInputs, !manifestRegenerationInputs.dirs.isEmpty {
+            stream <<< "nodes:\n"
+            for dir in manifestRegenerationInputs.dirs {
+                stream <<< "  " <<< Format.asJSON(dir) <<< ":\n"
+                stream <<< "    is-directory-structure: true\n"
+            }
+        }
+        
         stream <<< "commands: \n"
         for command in targets.allCommands.sorted(by: { $0.name < $1.name }) {
             stream <<< "  " <<< Format.asJSON(command.name) <<< ":\n"
             command.tool.append(to: stream)
             stream <<< "\n"
         }
+
+        if let manifestRegenerationInputs = manifestRegenerationInputs {
+            // Add command for computing manifest regeneration.
+            let regenerationCommand = ShellTool(
+                description: "",
+                inputs: manifestRegenerationInputs.dirs + manifestRegenerationInputs.files,
+                outputs: [plan.buildParameters.regenerateManifestToken.asString],
+                args: ["echo 1 >> " + plan.buildParameters.regenerateManifestToken.asString],
+                allowMissingInputs: true
+            )
+            stream <<< "  " <<< Format.asJSON(plan.buildParameters.regenerateManifestToken.asString) <<< ":\n"
+            regenerationCommand.append(to: stream)
+        }
+        
         try localFileSystem.writeFileContents(path, bytes: stream.bytes)
     }
+    
+    private func manifestRegenerationInputs() -> (dirs: [String], files: [String])? {
+        // If manifest caching is not enabled, just return nil from here.
+        guard plan.buildParameters.shouldEnableManifestCaching else { return nil }
+
+        var directoryNodesToTrack: [AbsolutePath] = []
+        var filesToTrack: [AbsolutePath] = []
+        
+        let graph = plan.graph
+        
+        for package in graph.packages {
+            // Track the package manifest.
+            // FIXME: We also need to add the resolved file here.
```

Comment:
Any reason not to just go ahead and fix this?

Reply:
No, just landing things incrementally.